packages/grpc-js/src/server-call.ts

```diff
@@ -553,105 +553,97 @@ export class Http2ServerCallStream<
-    const call = this;
-    const body: Buffer[] = [];
-    const limit = this.maxReceiveMessageSize;
+      function onData(chunk: Buffer) {
+        receivedLength += chunk.byteLength;
 
-    stream.on('data', onData);
-    stream.on('end', onEnd);
-    stream.on('error', onEnd);
+        if (limit !== -1 && receivedLength > limit) {
+          stream.removeListener('data', onData);
+          stream.removeListener('end', onEnd);
+          stream.removeListener('error', onEnd);
+
+          reject({
+            code: Status.RESOURCE_EXHAUSTED,
+            details: `Received message larger than max (${receivedLength} vs. ${limit})`,
+          });
+          return;
+        }
 
-    function onData(chunk: Buffer) {
-      receivedLength += chunk.byteLength;
+        body.push(chunk);
+      }
 
-      if (limit !== -1 && receivedLength > limit) {
+      function onEnd(err?: Error) {
         stream.removeListener('data', onData);
         stream.removeListener('end', onEnd);
         stream.removeListener('error', onEnd);
-        next({
-          code: Status.RESOURCE_EXHAUSTED,
-          details: `Received message larger than max (${receivedLength} vs. ${limit})`,
-        });
-        return;
-      }
 
-      body.push(chunk);
-    }
-
-    function onEnd(err?: Error) {
-      stream.removeListener('data', onData);
-      stream.removeListener('end', onEnd);
-      stream.removeListener('error', onEnd);
+        if (err !== undefined) {
+          reject({ code: Status.INTERNAL, details: err.message });
+          return;
+        }
 
-      if (err !== undefined) {
-        next({ code: Status.INTERNAL, details: err.message });
-        return;
-      }
+        if (receivedLength === 0) {
+          reject({
+            code: Status.INTERNAL,
+            details: 'received empty unary message',
+          });
+          return;
+        }
 
-      if (receivedLength === 0) {
-        next({
-          code: Status.INTERNAL,
-          details: 'received empty unary message',
-        });
-        return;
-      }
+        call.emit('receiveMessage');
 
-      call.emit('receiveMessage');
+        const requestBytes = Buffer.concat(body, receivedLength);
+        const compressed = requestBytes.readUInt8(0) === 1;
+        const compressedMessageEncoding = compressed ? encoding : 'identity';
+        const decompressedMessage = call.getDecompressedMessage(
+          requestBytes,
+          compressedMessageEncoding
+        );
 
-      const requestBytes = Buffer.concat(body, receivedLength);
-      const compressed = requestBytes.readUInt8(0) === 1;
-      const compressedMessageEncoding = compressed ? encoding : 'identity';
-      const decompressedMessage = call.getDecompressedMessage(
-        requestBytes,
-        compressedMessageEncoding
-      );
+        if (Buffer.isBuffer(decompressedMessage)) {
+          call.deserializeMessageWithInternalError(decompressedMessage);
```

Comment:
Now that this method returns the value, you have to do something with that value here and at the other call site. The way it currently works, I think you need another try/catch block here to pass the result into `resolve` or `reject` as appropriate. However, I think if you make `deserializeMessageWithInternalError` into an `async` function, it will automatically translate the `throw` there into a rejected promise. Then, since `resolve` here accepts a `PromiseLike<RequestType | void>`, you can pass the result of that call directly to `resolve` and it will handle both cases.

Reply:
Oh yeah you're right, I neglected the resolve there. I now turned it into an async function and specifically specified the resolve and catch at the call accordingly to make it very clear what is happening to the result. Otherwise it might become unclear to some, why `deserializeMessageWithInternalError` is async.

I thought if awaiting it might matter, because it would make the handling more complicated. But I don't think it makes a difference, because we're inside of another promise that only resolves after this call is finished anyway.